### PR TITLE
[Remote Test] Fix remote-run to process environment variables.

### PR DESCRIPTION
**Pull request by al45tair**
`remote-run` should look in the environment for input/output paths as well as considering command line arguments.

With this change, `test/Runtime/Paths.cpp` should work for remote testing and device testing.

rdar://106294557

### CODE REVIEW

**Comment by al45tair** [comment]
@swift-ci Please smoke test

File: utils/remote-run [hunk 1]
```diff
@@ -59,31 +59,111 @@ class CommandRunner(object):
...
-        sources = '\n'.join(sources)
-        if self.verbose:
-            print(sources, file=sys.stderr)
-        _, _ = rsync_proc.communicate(sources.encode('utf-8'))
-        if not self.ignore_rsync_failure and rsync_proc.returncode:
+        attempt = 1
+        rsync_proc = None
+
+        while attempt <= 3:
+            rsync_proc = self.popen(
```
Comment (compnerd) [inline utils/remote-run:129]: This feels like it might make sense to refactor and share so we have a retryable command.

File: utils/remote-run [hunk 2]
```diff
@@ -177,54 +267,85 @@ class ArgumentProcessor(object):
             while self.output_prefix.endswith(posixpath.sep):
                 self.output_prefix = self.output_prefix[:-len(posixpath.sep)]
 
+        if self.input_prefix:
+            split = posixpath.split(self.input_prefix)
```
Comment (compnerd) [inline utils/remote-run:271]: This might require some fixing in the future if we end up supporting remote run on Windows (which would be nice).

**Review by compnerd — APPROVED** [review]
Would be nice to get a second pair of eyes on this as well.  The shuffling around of the code made this a bit of a challenge to review very carefully.

File: utils/remote-run (hunk 1 above)
Comment (al45tair) [reply utils/remote-run:129]: These particular bits aren't really my changes; we had some of these internally and I thought it best to make the public version match the internal version here.

File: utils/remote-run (hunk 2 above)
Comment (al45tair) [reply utils/remote-run:271]: Agreed. Which might be doable these days given that Windows has an SSH server that (sort of mostly) works.

**Review by mikeash — APPROVED** [review]

**Pull request merged**